### PR TITLE
Add `expose-ids` feature

### PR DESCRIPTION
**Pull request by TheOnlyMrCat**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [ ] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
Discussed in matrix

**Description**
Adds the `expose-ids` feature, which allows API users to compare and store the identity of GPU objects through an opaque `Id` type. On native, `Id` wraps an unzipped `wgc::TypedId`; on web, `Id` wraps a manually-generated unique `u64`.

Most types with an `id` field now expose a feature-gated `global_id` method.

**Testing**
Not functionally tested yet, only typechecked

**Unresolved Questions**
- Should `Id`s be type-tagged, and if so, should there still be an untyped option?
- Should global atomics be used to guarantee uniqueness between multiple `Context`s on native?

### CODE REVIEW

**Comment by i509VCB** [comment]
Are these changes related to #3051?

File: wgpu/src/backend/web.rs [hunk 1]
```diff
@@ -903,22 +963,45 @@ fn map_map_mode(mode: crate::MapMode) -> u32 {
...
 
-fn future_request_adapter(result: JsFutureResult) -> Option<Sendable<web_sys::GpuAdapter>> {
+#[cfg(not(feature = "expose-ids"))]
+fn create_identified<T>(value: T, _id: u64) -> Identified<T> {
+    Identified(value)
+}
+
+#[cfg(feature = "expose-ids")]
+fn create_identified<T>(value: T, id: u64) -> Identified<T> {
+    Identified(value, id)
```
[the remark this answers is not in the document]
Comment (TheOnlyMrCat) [reply wgpu/src/backend/web.rs:973]: Oh! I didn't know those traits existed. Indexes into an array sounds like those ids would be reused on deallocation (and that's what the example code in [the `wasm-bindgen` guide](https://rustwasm.github.io/docs/wasm-bindgen/contributing/design/js-objects-in-rust.html#long-lived-js-objects) looks like it's doing), and I don't think that's appropriate for this use-case.

**Comment by cwfitzgerald** [comment]
@i509VCB this particular use case is for unique identification of various resources for use in a cache, they need to be globally unique, not just unique-at-any-one-time.

File: wgpu/src/backend/web.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (i509VCB) [reply wgpu/src/backend/web.rs:973]: If the requirement is unique, wouldn't the wgpu-core tracker reusing old IDs cause a conflict?

[the remark this answers is not in the document]
Comment (TheOnlyMrCat) [reply wgpu/src/backend/web.rs:973]: It reuses freed indices, but also keeps track of the "epoch" of those indices. Reusing them still creates a unique id.

[the remark this answers is not in the document]
Comment (ErichDonGubler) [reply wgpu/src/backend/web.rs:973]: @TheOnlyMrCat: Ah, so, they're [generational indices](https://lucassardois.medium.com/generational-indices-guide-8e3c5f7fd594)?

File: wgpu/src/backend/web.rs [hunk 2]
```diff
@@ -1,6 +1,8 @@
 #![allow(clippy::type_complexity)]
 
 use js_sys::Promise;
+#[cfg(feature = "expose-ids")]
+use std::sync::atomic::{self, AtomicU64};
```
Comment (cwfitzgerald) [inline wgpu/src/backend/web.rs:5]: For cfg stuff, lets just fully qualify

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Looks great outside of a small comment

**Comment by cwfitzgerald** [comment]
Sorry for the long delay in reviewing

File: wgpu/src/backend/web.rs (hunk 2 above)
Comment (TheOnlyMrCat) [reply wgpu/src/backend/web.rs]: Done!

**Review by cwfitzgerald — APPROVED** [review]
Thanks!

**Pull request merged**